### DISCUSSION TITLE
Auto mod and Cinema mod enable at the same time

### DISCUSSION THREAD

**Discussion by cdwcgt** [ppy/osu#16018] [Q&A] [answered]
I enabled cinema mode and use Ctrl + enter to play. Because shortcut key so Autoplay also enabled
![image](https://user-images.githubusercontent.com/34775378/145376649-c42d85ca-f340-48c5-809b-8c8b60b9c284.png)
But Autoplay should not actually be enabled (Because stable won't enable autoplay)
is it a bug?

**Comment by peppy** [comment]
Does indeed seem like unintended behaviour. Tried to fix with the simplest of fixes but it doesn't work so great.

```diff

diff --git a/osu.Game/Rulesets/Mods/ModAutoplay.cs b/osu.Game/Rulesets/Mods/ModAutoplay.cs
index 60b9c29fe0..2debc11f43 100644
--- a/osu.Game/Rulesets/Mods/ModAutoplay.cs
+++ b/osu.Game/Rulesets/Mods/ModAutoplay.cs
@@ -26,7 +26,7 @@ public abstract class ModAutoplay : Mod, IApplicableFailOverride, ICreateReplay
 
         public override bool UserPlayable => false;
 
-        public override Type[] IncompatibleMods => new[] { typeof(ModRelax), typeof(ModFailCondition), typeof(ModNoFail) };
+        public override Type[] IncompatibleMods => new[] { typeof(ModRelax), typeof(ModFailCondition), typeof(ModNoFail), typeof(ModAutoplay) };
 
         public override bool HasImplementation => GetType().GenericTypeArguments.Length == 0;
 

```

**Reply by frenzibyte** [reply]
Normally when a new mod is added to the selected list it should deselect whichever mod incompatible in that list, but since it is expected here that `SongSelect` should instead not append `ModAutoplay` when `ModCinema` exists, maybe it would make more sense to handle it there instead?:
```diff
diff --git a/osu.Game/Screens/Select/PlaySongSelect.cs b/osu.Game/Screens/Select/PlaySongSelect.cs
index 94aa165785..b7207ef2c6 100644
--- a/osu.Game/Screens/Select/PlaySongSelect.cs
+++ b/osu.Game/Screens/Select/PlaySongSelect.cs
@@ -14,6 +14,7 @@
 using osu.Game.Screens.Play;
 using osu.Game.Screens.Ranking;
 using osu.Game.Users;
+using osu.Game.Utils;
 using osuTK.Input;

 namespace osu.Game.Screens.Select
@@ -95,11 +96,11 @@ protected override bool OnStart()
                     return false;
                 }

-                var mods = Mods.Value;
+                var newMods = Mods.Value.Append(autoInstance).ToArray();

-                if (mods.All(m => m.GetType() != autoInstance.GetType()))
+                if (ModUtils.CheckCompatibleSet(newMods))
                 {
-                    Mods.Value = mods.Append(autoInstance).ToArray();
+                    Mods.Value = newMods;
                     removeAutoModOnResume = true;
                 }
             }
```

But it actually appears that `ModAutoplay` and `ModCinema` aren't incompatible with each other, and worser than that, `ModCinema` inherits from `ModAutoplay`.

I've looked at separating the code shared between both mods and the results of that as can be seen in https://github.com/ppy/osu/compare/master..frenzibyte:autoplay-cinema-incompatibility. Does that sound like a good direction?

An alternative direction would be to inline the code in the abstract mod class to both `ModAutoplay` and `ModCinema`.

**Reply by peppy** [reply]
No idea. Do mods not support listing themselves as incompatible in abstract classes?

**Reply by frenzibyte** [reply]
Right, marking `ModAutoplay` and `ModCinema` as incompatible directly without structural changes works as well given the robustness of `ModUtils`: https://github.com/ppy/osu/compare/master..frenzibyte:autoplay-cinema-incompatibility-alt.

**Comment by frenzibyte** [comment, marked as answer]
Tracking in #17508
